lgc/util/CacheHashNoteUtils.cpp

```diff
@@ -0,0 +1,298 @@
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+#include "lgc/util/CacheHashNoteUtils.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Object/ELF.h"
+#include "llvm/Support/Alignment.h"
+#include "llvm/Support/BinaryByteStream.h"
+#include "llvm/Support/BinaryStreamReader.h"
+#include "llvm/Support/BinaryStreamWriter.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace lgc;
+
+// The implementation of ELF rewriting is based on "Linux Programmer's Manual ELF(5)".
+// In particular, see "Notes (Nhdr)" of the document for the note section.
+
+namespace {
+
+// It is similar to struct NoteHeader defined in llpc/util/vkgcElfReader.h, but
+// it allows us to use Elf_Nhdr_Impl<object::ELF64LE>::Align and it does not
+// have the size limitation of note name.
+using NoteHeader = object::Elf_Nhdr_Impl<object::ELF64LE>;
+
+// =====================================================================================================================
+// A section to shift and its new offset.
+struct SectionShiftInfo {
+  SmallString<64> section;
+  ELF::Elf64_Off newOffset;
+};
+
+// =====================================================================================================================
+// Create a note section with the note "llpc_cache_hash" that has the cache hash
+// as the description.
+//
+// Reference: Linux Programmer's Manual ELF(5) "Notes (Nhdr)".
+//
+// @param cacheHash               : Hash code to be added to the new .note section.
+// @param llpcVersion             : LLPC version that gives us a hint how cacheHash
+//                                was generated.
+// @param [out] noteSectionStream : The new .note section with the cache hash.
+void createNoteSectionWithCacheHash(StringRef cacheHash, StringRef llpcVersion,
+                                    AppendingBinaryByteStream &noteSectionStream) {
+  BinaryStreamWriter noteSectionWriter(noteSectionStream);
+
+  // Write the note header for the cache hash.
+  NoteHeader noteHeader;
+  noteHeader.n_namesz = sizeof(CacheHashNoteName);
+  noteHeader.n_descsz = cacheHash.size();
+  // TODO: Define a note type to specify cache hash. It must be defined in
+  // llvm/include/llvm/BinaryFormat/ELF.h.
+  // Note types with values between 0 and 32 (inclusive) are reserved.
+  noteHeader.n_type = 0;
+  auto error =
+      noteSectionWriter.writeFixedString(StringRef(reinterpret_cast<const char *>(&noteHeader), sizeof(noteHeader)));
+  assert(!error);
+
+  // Write the cache hash note name terminated by zero and zeros for the alignment.
+  error = noteSectionWriter.writeFixedString(StringRef(CacheHashNoteName, sizeof(CacheHashNoteName)));
```

Comment:
If you want to write the string without the null terminator, you can use: `noteSectionWriter.writeFixedString(CacheHashNoteName)`. If you also want the null terminator, you can use `noteSectionWriter.writeCString(CacheHashNoteName)`.
See https://llvm.org/doxygen/classllvm_1_1BinaryStreamWriter.html#a07608f1ac2a8045b1b72108b840a8ca3